qutip/models/piqs.py

```diff
@@ -0,0 +1,1201 @@
+        -------
+        liouv: :class: qutip.Qobj
+            The Liouvillian matrix for the system.
+        """
+        lindblad = self.lindbladian()
+        if self.hamiltonian is None:
+            liouv = lindblad
+
+        else:
+            hamiltonian = self.hamiltonian
+            hamiltonian_superoperator = - 1j * \
+                spre(hamiltonian) + 1j * spost(hamiltonian)
+            liouv = lindblad + hamiltonian_superoperator
+        return liouv
+
+    def eigenstates(self, liouvillian):
+        """Calculate the eigenvalues and eigenvectors of the Liouvillian.
+
+        It removes the spurious eigenvectors with corresponding
+        non-hermitian density matrices.
+
+        Parameters
+        ----------
+        liouvillian: :class: qutip.Qobj
+            The Liouvillian of which to calculate the spectrum.
+
+        Returns
+        -------
+        eigen_states: list
+            The list of eigenvalues and correspondent eigenstates.
+        """
+        unpruned_eigenstates = liouvillian.eigenstates()
+        eigen_states = self.prune_eigenstates(unpruned_eigenstates)
+        return eigen_states
+
+    def prune_eigenstates(self, liouvillian_eigenstates):
+        """Remove spurious eigenvalues and eigenvectors of the Liouvillian.
+
+        Spurious means that the given eigenvector has elements outside of the
+        block-diagonal matrix.
+
+        Parameters
+        ----------
+        liouvillian_eigenstates: list
+            A list with the eigenvalues and eigenvectors of the Liouvillian
+            including spurious ones.
+
+        Returns
+        -------
+        correct_eigenstates: list
+            The list with the correct eigenvalues and eigenvectors of the
+            Liouvillian.
+        """
+        N = self.N
+        block_mat = block_matrix(N)
+        nnz_tuple_bm = [(i, j) for i, j in zip(*block_mat.nonzero())]
+
+        # 0. Create  a copy of the eigenvalues to approximate values
+        eig_val, eig_vec = liouvillian_eigenstates
+        tol = 10
+        eig_val_round = np.round(eig_val, tol)
+
+        # 2. Use 'block_matrix(N)' to remove eigenvectors with matrix
+        # elements
+        # outside of the block matrix.
+        forbidden_eig_index = []
+        for k in range(0, len(eig_vec)):
+            dm = vector_to_operator(eig_vec[k])
+            nnz_tuple = [(i, j) for i, j in zip(*dm.data.nonzero())]
+            for i in nnz_tuple:
+                if i not in nnz_tuple_bm:
+                    if np.round(dm[i], tol) != 0:
+                        forbidden_eig_index.append(k)
+
+        forbidden_eig_index = np.array(list(set(forbidden_eig_index)))
+        # 3. Remove the forbidden eigenvalues and eigenvectors.
+        correct_eig_val = np.delete(eig_val, forbidden_eig_index)
+        correct_eig_vec = np.delete(eig_vec, forbidden_eig_index)
+        correct_eigenstates = correct_eig_val, correct_eig_vec
+        return correct_eigenstates
+
+
```

Comment:
single newline between functions

Reply:
Is there a single newline here too? I thought that classes are wrapped with two newlines? pep8 linter didn't catch this though. I 'll leave it for you to decide